File: frontend/src/services/userService.js
import api from './api';

export const userService = {
  // Получить всех пользователей
  getAllUsers: async () => {
    const response = await api.get('/users');
    return response.data;
  },

  // Получить пользователя по ID
  getUserById: async (id) => {
    const response = await api.get(`/users/${id}`);
    return response.data;
  },

  // Обновить пользователя
  updateUser: async (id, userData) => {
    const response = await api.put(`/users/${id}`, userData);
    return response.data;
  },

  // Удалить пользователя
  deleteUser: async (id) => {
    await api.delete(`/users/${id}`);
  },

  // Изменить роль пользователя
  changeUserRole: async (id, newRole) => {
    const user = await userService.getUserById(id);
    const updatedUser = {
      ...user,
      role: newRole
    };
    return await userService.updateUser(id, updatedUser);
  },

  // Получить пользователя по email
  getUserByEmail: async (email) => {
    const response = await api.get(`/users/email/${email}`);
    return response.data;
  },

  // Получить количество пользователей
  getUserCount: async () => {
    const response = await api.get('/users/count');
    return response.data;
  }
};

File: src/main/java/com/example/culturaleventssystem/service/UserService.java
package com.example.culturaleventssystem.service;

import com.example.culturaleventssystem.entity.User;
import com.example.culturaleventssystem.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    // Создать нового пользователя
    public User createUser(User user) {
        // Проверяем, нет ли пользователя с таким email или username
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new RuntimeException("User with this email already exists");
        }
        if (userRepository.existsByUsername(user.getUsername())) {
            throw new RuntimeException("User with this username already exists");
        }
        return userRepository.save(user);
    }

    // Получить всех пользователей
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // Найти пользователя по ID
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

    // Найти пользователя по email
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    // Найти пользователя по username
    public Optional<User> getUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }

    // Обновить пользователя
    public User updateUser(Long id, User userDetails) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));

        // Проверяем уникальность email, если он изменен
        if (!user.getEmail().equals(userDetails.getEmail()) &&
                userRepository.existsByEmail(userDetails.getEmail())) {
            throw new RuntimeException("User with this email already exists");
        }

        // Проверяем уникальность username, если он изменен
        if (!user.getUsername().equals(userDetails.getUsername()) &&
                userRepository.existsByUsername(userDetails.getUsername())) {
            throw new RuntimeException("User with this username already exists");
        }

        user.setUsername(userDetails.getUsername());
        user.setEmail(userDetails.getEmail());
        user.setPassword(userDetails.getPassword());
        user.setRole(userDetails.getRole());

        return userRepository.save(user);
    }

    // Удалить пользователя
    public void deleteUser(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found with id: " + id));
        userRepository.delete(user);
    }

    // Аутентификация пользователя
    public Optional<User> authenticateUser(String email, String password) {
        return userRepository.findByEmailAndPassword(email, password);
    }

    // Проверить существование пользователя по email
    public boolean userExistsByEmail(String email) {
        return userRepository.existsByEmail(email);
    }

    // Получить количество пользователей
    public long getUserCount() {
        return userRepository.count();
    }
}